File: apas-frontend/lib/encryption.ts
// lib/encryption.ts
import crypto from 'node:crypto';

const ENCRYPTION_PASSWORD = process.env.VAULT_ENCRYPTION_PASSWORD;
const TAG_LENGTH = 16;

function deriveKey(password: string): Buffer {
  // 🔍 Debug: log the derived key prefix (first 8 bytes in hex)
  const key = crypto.scryptSync(password, 'salt', 32, { N: 16384, r: 8, p: 1 });
  console.log(`[deriveKey] Derived key prefix: ${key.subarray(0, 8).toString('hex')}`);
  return key;
}

export function encrypt(text: string): string {
  if (!ENCRYPTION_PASSWORD) {
    throw new Error('VAULT_ENCRYPTION_PASSWORD is not set');
  }
  const key = deriveKey(ENCRYPTION_PASSWORD);
  const iv = crypto.randomBytes(16);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  let encrypted = cipher.update(text, 'utf8', 'hex');
  encrypted += cipher.final('hex');
  const tag = cipher.getAuthTag();
  return `enc:${iv.toString('hex')}:${tag.toString('hex')}:${encrypted}`;
}

export function decrypt(encryptedText: string): string {
  if (!ENCRYPTION_PASSWORD) {
    throw new Error('VAULT_ENCRYPTION_PASSWORD is not set');
  }
  if (!encryptedText.startsWith('enc:')) return encryptedText;

  // 🔍 Debug: log the password prefix
  console.log(`[decrypt] Password (first 4 chars): ${ENCRYPTION_PASSWORD.substring(0, 4)}`);

  const payload = encryptedText.substring(4);
  const parts = payload.split(':');

  // --- Python format (2 parts): nonce_base64 : ct_base64 (ciphertext + tag) ---
  if (parts.length === 2) {
    const [nonceB64, ctB64] = parts;
    const nonce = Buffer.from(nonceB64, 'base64');
    const ct = Buffer.from(ctB64, 'base64');

    console.log('[decrypt] Python format:');
    console.log(`  nonce length: ${nonce.length} bytes`);
    console.log(`  ct length: ${ct.length} bytes`);

    // The tag is the last 16 bytes
    if (ct.length < TAG_LENGTH) {
      throw new Error('Ciphertext too short to contain tag');
    }
    const tag = ct.subarray(-TAG_LENGTH);
    const ciphertext = ct.subarray(0, -TAG_LENGTH);

    console.log(`  tag length: ${tag.length} bytes`);
    console.log(`  ciphertext length: ${ciphertext.length} bytes`);

    const key = deriveKey(ENCRYPTION_PASSWORD);
    const decipher = crypto.createDecipheriv('aes-256-gcm', key, nonce);
    decipher.setAuthTag(tag);

    const decryptedBuffer = Buffer.concat([
      decipher.update(ciphertext),
      decipher.final()
    ]);
    return decryptedBuffer.toString('utf8');
  }

  // --- Node format (3 parts): ivHex : tagHex : encryptedHex ---
  if (parts.length === 3) {
    const [ivHex, tagHex, encryptedHex] = parts;
    const iv = Buffer.from(ivHex, 'hex');
    const tag = Buffer.from(tagHex, 'hex');
    const encrypted = Buffer.from(encryptedHex, 'hex');
    const key = deriveKey(ENCRYPTION_PASSWORD);

    const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
    decipher.setAuthTag(tag);

    const decryptedBuffer = Buffer.concat([
      decipher.update(encrypted),
      decipher.final()
    ]);
    return decryptedBuffer.toString('utf8');
  }

  throw new Error(`Invalid encrypted format: ${parts.length} parts`);
}

File: apas-frontend/lib/encryption.py
import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt
from cryptography.hazmat.backends import default_backend
from .config import config
from .logger import logger

ENCRYPTION_PASSWORD = config.encryption.password

if not ENCRYPTION_PASSWORD:
    logger.warning('VAULT_ENCRYPTION_PASSWORD not set; vault will be stored in plain text.')

def _derive_key(password, salt=b'salt'):
    kdf = Scrypt(
        salt=salt,
        length=32,
        n=2**14,
        r=8,
        p=1,
        backend=default_backend()
    )
    return kdf.derive(password.encode('utf-8'))

def encrypt(text):
    if not ENCRYPTION_PASSWORD:
        return f"plain:{text}"
    key = _derive_key(ENCRYPTION_PASSWORD)
    aesgcm = AESGCM(key)
    nonce = os.urandom(12)
    ct = aesgcm.encrypt(nonce, text.encode('utf-8'), None)
    # Combine nonce + ciphertext
    return f"enc:{base64.b64encode(nonce).decode('utf-8')}:{base64.b64encode(ct).decode('utf-8')}"

def decrypt(encrypted_text):
    if not ENCRYPTION_PASSWORD:
        if encrypted_text.startswith('plain:'):
            return encrypted_text[6:]
        return encrypted_text
    if encrypted_text.startswith('plain:'):
        return encrypted_text[6:]
    if not encrypted_text.startswith('enc:'):
        return encrypted_text
    parts = encrypted_text[4:].split(':')
    if len(parts) != 2:
        raise ValueError('Invalid encrypted format')
    nonce_b64, ct_b64 = parts
    nonce = base64.b64decode(nonce_b64)
    ct = base64.b64decode(ct_b64)
    key = _derive_key(ENCRYPTION_PASSWORD)
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(nonce, ct, None)
    return plaintext.decode('utf-8')